// @tarojs/taro-rn/dist/lib/showActionSheet/index.js
import React from 'react';
import RootSiblings from 'react-native-root-siblings';
import ActionSheet from './ActionSheet';
function showActionSheet(obj) {
    const { itemList = [], itemColor = '#000000', success, fail, complete } = obj || {};
    const autoDectect = true;
    const type = 'ios';
    return new Promise((resolve, reject) => {
        let sibling;
        function onSuccess(tapIndex) {
            sibling && sibling.destroy();
            sibling = undefined;
            const res = { tapIndex, errMsg: 'showActionSheet:ok' };
            // fix: iOS 无法打开相册
            // https://github.com/expo/expo/issues/25705
            setTimeout(() => {
                success === null || success === void 0 ? void 0 : success(res);
                complete === null || complete === void 0 ? void 0 : complete(res);
                resolve(res);
            }, 1);
        }
        function onFail() {
            const res = { errMsg: 'showActionSheet:fail cancel' };
            sibling && sibling.destroy();
            sibling = undefined;
            fail === null || fail === void 0 ? void 0 : fail(res);
            complete === null || complete === void 0 ? void 0 : complete(res);
            reject(res);
        }
        sibling = new RootSiblings(React.createElement(ActionSheet, { autoDectect: autoDectect, type: type, visible: true, onClose: onFail, menus: itemList.map((item, index) => {
                return {
                    type: 'default',
                    label: item,
                    textStyle: { color: itemColor },
                    onPress: () => onSuccess(index)
                };
            }), actions: [
                {
                    type: 'default',
                    label: '取消',
                    textStyle: { color: itemColor },
                    onPress: onFail
                }
            ] }));
    });
}
export { showActionSheet };
